### PR TITLE
created directory tree, fixed warnings, cmake improvements

### PR DESCRIPTION
**Pull request by bashbaug**
This change:

* Moves the ICD loader files into a "loader" directory, with sub-directories for Windows and Linux.
* Renames the "platform" directory to "log", since it contains logging functionality.
* Adds a few CMakefile improvements, such as a cache variable for the OpenCL headers, and the define for CL_TARGET_OPENCL_VERSION.
* Added include files and missing declarations to fix warnings.

Tested both Windows and Linux.

### CODE REVIEW

**Comment by bashbaug** [comment]
Thank you for a very thorough review!

I've addressed a number of your comments, and I've tested my changes on an old Ubuntu 14.04 VM with CMake 2.8.12, so we're good at least back that far.  I agree we could do more if we wanted to use a newer version of CMake, but I think this fixes most of the big CMake issues.  Would you mind accepting this PR as-is and filing an github issue to address the remaining issues?

File: test/log/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+add_library (IcdLog SHARED icd_test_log.c)
+if (WIN32)
+    target_compile_definitions (IcdLog PRIVATE _CRT_SECURE_NO_WARNINGS)
```
[the remark this answers is not in the document]
Comment (bashbaug) [reply test/log/CMakeLists.txt]: Fixed - I think this was introduced in CMake 2.8.11.  I've changed the minimum required version to 2.8.11 and I've tested on a system with 2.8.12.

File: CMakeLists.txt [hunk 2]
```diff
@@ -35,22 +35,34 @@ option (BUILD_SHARED_LIBS "Build shared libs" ON)
 set (CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
 set (CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
 
-set (OPENCL_ICD_LOADER_SOURCES icd.c icd_dispatch.c)
+set (OPENCL_ICD_LOADER_SOURCES
```
[the remark this answers is not in the document]
Comment (bashbaug) [reply CMakeLists.txt]: I haven't made this change yet.

File: CMakeLists.txt [hunk 3]
```diff
@@ -61,7 +73,7 @@ else()
     if (APPLE)
         set_target_properties (OpenCL PROPERTIES LINK_FLAGS "-pthread")
```
[the remark this answers is not in the document]
Comment (bashbaug) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt [hunk 4]
```diff
@@ -35,22 +35,34 @@ option (BUILD_SHARED_LIBS "Build shared libs" ON)
 set (CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
```
[the remark this answers is not in the document]
Comment (bashbaug) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt [hunk 5]
```diff
@@ -35,22 +35,34 @@ option (BUILD_SHARED_LIBS "Build shared libs" ON)
...
-endif ()
+set (OPENCL_ICD_LOADER_HEADERS_DIR ${CMAKE_CURRENT_SOURCE_DIR}/inc CACHE PATH "Path to OpenCL Headers")
+
+include_directories (${OPENCL_ICD_LOADER_HEADERS_DIR})
+include_directories (loader)
 
-include_directories (${OPENCL_INCLUDE_DIRS})
+add_definitions (-DCL_TARGET_OPENCL_VERSION=220)
 
 add_library (OpenCL SHARED ${OPENCL_ICD_LOADER_SOURCES})
```
[the remark this answers is not in the document]
Comment (bashbaug) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt [hunk 6]
```diff
@@ -67,13 +72,23 @@ if (WIN32)
...
     endif ()
 endif ()
 
+include_directories (${OPENCL_ICD_LOADER_HEADERS_DIR})
+add_definitions (-DCL_TARGET_OPENCL_VERSION=220)
+
+target_include_directories (OpenCL PRIVATE loader)
 target_link_libraries (OpenCL ${CMAKE_DL_LIBS})
 
 enable_testing()
```
Comment (mmha) [inline CMakeLists.txt]: Could you also `include(CTest)` and wrap the tests with `BUILD_TESTING`?

```cmake
include(CTest) # Adds the BUILD_TESTING option (on by default) and `enable_testing()` if testing is enabled
# ...
if(BUILD_TESTING)
    add_subdirectory(test)
endif()
```

**Comment by mmha** [comment]
Thanks for updating it!

Just this small comment, I guess the rest boils down to which CMake version we want to require. So I'm okay with merging it as-is and I'll create a separate issue/PR later.

File: CMakeLists.txt (hunk 6 above)
Comment (bashbaug) [reply CMakeLists.txt]: Fixed - that's a CMake trick I hadn't seen before.  Thanks!

**Comment by bashbaug** [comment]
Great.  I think this is ready to go now.

**Comment by mmha** [comment]
Yes, I think it can be merged now.

**Comment by kepatil** [comment]
I made a cursory check on the change. Looks good enough to merge based on comments above.

**Pull request merged**